### PR TITLE
Improve/fix response status code handling

### PR DESCRIPTION
**Pull request by sparrc**
###### Required for all non-trivial PRs
- [ ] Rebased/mergable
- [ ] CHANGELOG.md updated

see https://github.com/influxdata/telegraf/issues/1331

### CODE REVIEW

**Comment by sparrc** [comment]
Changing client behavior to check status code immediately after making an HTTP request, rather than doing complicated if-statements after decoding a response that can't possibly be JSON. Also makes the errors a little easier to read, from this:

```
unable to decode json: received status code 404 err: json: cannot unmarshal number into Go value of type client.Response
```

to this:

```
Received bad status code from server: 404 (404 Not Found)
```

File: client/v2/client.go [hunk 1]
```diff
@@ -565,24 +565,29 @@ func (c *client) Query(q Query) (*Response, error) {
...
+	if resp.StatusCode >= 400 {
+		return nil, fmt.Errorf("Received bad status code from server: %d (%s)",
+			resp.StatusCode, resp.Status)
+	}
+
 	var response Response
 	dec := json.NewDecoder(resp.Body)
 	dec.UseNumber()
-	decErr := dec.Decode(&response)
+	err = dec.Decode(&response)
```
Comment (benbjohnson) [inline client/v2/client.go]: You can merge this with the `if` below:

```go
if err := dec.Decode(&response); err != nil {
```
```

File: client/v2/client.go [hunk 2]
```diff
@@ -565,24 +565,29 @@ func (c *client) Query(q Query) (*Response, error) {
 	}
 	defer resp.Body.Close()
 
+	// check status code before doing any JSON marshalling.
+	// if the status code is >= 400, there won't be any valid json to marshal so
+	// return immediately.
+	if resp.StatusCode >= 400 {
+		return nil, fmt.Errorf("Received bad status code from server: %d (%s)",
+			resp.StatusCode, resp.Status)
```
Comment (benbjohnson) [inline client/v2/client.go]: For consistency, can you change the initial letter to lowercase? Also on the other `Errorf()` calls below.

**Comment by jsternberg** [comment]
@sparrc I think the reason we do that logic is because JSON actually does get returned on a 500 error.

https://github.com/influxdata/influxdb/blob/a27f2a3cb08774792d96b8f806d871b24db17a86/services/httpd/handler.go#L647

**Comment by sparrc** [comment]
Ah, my bad, we can close then.

**Pull request closed without merging**